File: src/solar/solar-common/w3ts/handles/player.ts
/** @noSelfInFile **/

import {Force} from "./force";
import {Handle} from "./handle";
import {Point} from "./point";
import DataBase from "../../common/DataBase";

const cache: { [phid: number]: MapPlayer } = {}

export class MapPlayer extends Handle<player> {

    //GetUserId 对战平台的账号id
    userId?: number;

    private constructor(index: number) {
        if (Handle.initFromHandle()) {
            super();
        } else {
            super(Player(index));
        }
    }

    /**
     * 太阳 自定义数据集
     */
    public get solarData(): AppPlayerData {
        return DataBase.getPlayerSolarData(this.handle)
    }

    public set solarData(obj: { [key: string]: any }) {
        DataBase.setDataByHandle("+ply", this.handle, obj)
    }

    /**
     * 本地玩家 [R]
     */
    public isLocalPlayer() {
        if (GetLocalPlayer() == this.handle) {
            return true
        }
        return false;
    }

    /**
     * 获取玩家所有单位 （新建单位组）
     */
    public getAllUnits(): group {
        let g = CreateGroup()
        GroupEnumUnitsOfPlayer(g, this.handle, null)
        return g;
    }

    /**
     * 对玩家显示文本消息(指定时间) [R]
     */
    public displayText(text: string, duration: number = 5) {
        DisplayTimedTextToPlayer(this.handle, 0, 0, duration, text);
    }

    // /**
    //  * 同步获取当前玩家选择的第一个单位
    //  *
    //  * @deprecated see PlayerUtil.syncSelectedUnit
    //  */
    // public getSelectedUnit(): unit {
    //     SyncSelections()
    //     GroupClear(tempGroup)
    //     GroupEnumUnitsSelected(tempGroup, this.handle, null)
    //     let unitHandle = FirstOfGroup(tempGroup)
    //     return unitHandle;
    // }

    /**
     * 玩家颜色 [R]
     */
    public set color(color: playercolor) {
        SetPlayerColor(this.handle, color);
    }

    /**
     * 玩家颜色 [R]
     */
    public get color() {
        return GetPlayerColor(this.handle);
    }

    /**
     * 玩家控制者
     */
    public get controller() {
        return GetPlayerController(this.handle);
    }

    /**经验上限*/
    public get handicap() {
        return GetPlayerHandicap(this.handle);
    }

    /**
     * 设置经验上限 [R]
     */
    public set handicap(handicap: number) {
        SetPlayerHandicap(this.handle, handicap);
    }

    /***
     * 经验获得率
     */
    public get handicapXp() {
        return GetPlayerHandicapXP(this.handle);
    }

    /**
     * 设置经验获得率 [R]
     */
    public set handicapXp(handicap: number) {
        SetPlayerHandicapXP(this.handle, handicap);
    }

    /**
     * 玩家ID - 1 [R]
     */
    public get id() {
        return GetPlayerId(this.handle);
    }

    /**
     * 玩家名字
     */
    public get name() {
        return GetPlayerName(this.handle);
    }

    /**
     * 设置玩家名字
     */
    public set name(value: string) {
        SetPlayerName(this.handle, value);
    }

    /**
     * 玩家的种族
     */
    public get race() {
        return GetPlayerRace(this.handle);
    }

    /**
     * 玩家游戏属性
     */
    public get slotState() {
        return GetPlayerSlotState(this.handle);
    }

    /**
     * 玩家开始点
     */
    public get startLocation() {
        return GetPlayerStartLocation(this.handle);
    }

    /**
     * 起始点x
     */
    public get startLocationX() {
        return GetStartLocationX(this.startLocation);
    }

    /**
     * 起始点y
     */
    public get startLocationY() {
        return GetStartLocationY(this.startLocation);
    }

    /**
     * 起始点
     */
    public get startLocationPoint() {
        return GetStartLocationLoc(this.startLocation);
    }

    /**
     * 玩家所在的队伍
     */
    public get team() {
        return GetPlayerTeam(this.handle);
    }

    // public get townHallCount() {
    //   return BlzGetPlayerTownHallCount(this.handle);
    // }

    /**
     * 增加科技等级
     */
    public addTechResearched(techId: number | string, levels: number) {
        AddPlayerTechResearched(this.handle, techId, levels);
    }

    // public decTechResearched(techId: number | string, levels: number) {
    //   BlzDecPlayerTechResearched(this.handle, techId, levels);
    // }

    // Used to store hero level data for the scorescreen
    // before units are moved to neutral passive in melee games
    /**
     * 缓存玩家数据
     */
    public cacheHeroData() {
        CachePlayerHeroData(this.handle);
    }

    /**
     * 玩家与玩家结盟
     */
    public compareAlliance(otherPlayer: MapPlayer, whichAllianceSetting: alliancetype) {
        return GetPlayerAlliance(this.handle, otherPlayer.handle, whichAllianceSetting);
    }

    /**
     * 坐标在迷雾中
     */
    public coordsFogged(x: number, y: number) {
        return IsFoggedToPlayer(x, y, this.handle);
    }

    /**
     * 坐标在黑色阴影中
     */
    public coordsMasked(x: number, y: number) {
        return IsMaskedToPlayer(x, y, this.handle);
    }

    /**
     * 坐标可见
     */
    public coordsVisible(x: number, y: number) {
        return IsVisibleToPlayer(x, y, this.handle);
    }

    /**
     * 削弱玩家
     */
    public cripple(toWhichPlayers: Force, flag: boolean) {
        CripplePlayer(this.handle, toWhichPlayers.handle, flag);
    }

    /**
     * 获得玩家得分
     */
    public getScore(whichPlayerScore: playerscore) {
        return GetPlayerScore(this.handle, whichPlayerScore);
    }

    /**
     * 获得玩家属性
     */
    public getState(whichPlayerState: playerstate) {
        return GetPlayerState(this.handle, whichPlayerState);
    }

    /**
     * 获得建筑数量
     */
    public getStructureCount(includeIncomplete: boolean) {
        return GetPlayerStructureCount(this.handle, includeIncomplete);
    }

    /**
     * 玩家税率 [R]
     */
    public getTaxRate(otherPlayer: player, whichResource: playerstate) {
        return GetPlayerTaxRate(this.handle, otherPlayer, whichResource);
    }

    /**
     * 获取玩家科技数量
     */
    public getTechCount(techId: number | string, specificonly: boolean) {
        return GetPlayerTechCount(this.handle, techId, specificonly);
    }


    /**
     * 获取允许玩家的科技上限
     * 单位默认建造数量为2147483647 */
    public getTechMaxAllowed(techId: number | string) {
        return GetPlayerTechMaxAllowed(this.handle, techId);
    }

    /**
     * 获取玩家科技是否已经研究
     */
    public getTechResearched(techId: number | string, specificonly: boolean) {
        return GetPlayerTechResearched(this.handle, techId, specificonly);
    }

    /**
     * 单位数量
     */
    public getUnitCount(includeIncomplete: boolean) {
        return GetPlayerUnitCount(this.handle, includeIncomplete);
    }

    /**
     * 获取玩家特定单位数
     * @param unitName
     * @param includeIncomplete
     * @param includeUpgrades
     */
    public getUnitCountByType(unitName: string, includeIncomplete: boolean, includeUpgrades: boolean) {
        return GetPlayerTypedUnitCount(this.handle, unitName, includeIncomplete, includeUpgrades);
    }

    /**
     * 玩家在玩家组
     */
    public inForce(whichForce: Force) {
        return IsPlayerInForce(this.handle, whichForce.handle);
    }

    /**
     * 玩家是裁判或观察者 [R]
     */
    public isObserver() {
        return IsPlayerObserver(this.handle);
    }

    /**
     * 玩家是玩家的同盟
     */
    public isPlayerAlly(otherPlayer: MapPlayer) {
        return IsPlayerAlly(this.handle, otherPlayer.handle);
    }

    /**
     * 玩家是玩家的敌人
     */
    public isPlayerEnemy(otherPlayer: MapPlayer) {
        return IsPlayerEnemy(this.handle, otherPlayer.handle);
    }

    /**
     * 玩家的种族选择
     */
    public isRacePrefSet(pref: racepreference) {
        return IsPlayerRacePrefSet(this.handle, pref);
    }

    /**
     * 玩家是否可选
     */
    public isSelectable() {
        return GetPlayerSelectable(this.handle);
    }

    /**
     * 点被迷雾遮挡
     */
    public pointFogged(whichPoint: Point) {
        return IsLocationFoggedToPlayer(whichPoint.handle, this.handle);
    }

    /**
     * 点被黑色阴影遮挡
     */
    public pointMasked(whichPoint: Point) {
        return IsLocationMaskedToPlayer(whichPoint.handle, this.handle);
    }

    /**
     * 点对于玩家可见
     */
    public pointVisible(whichPoint: Point) {
        return IsLocationVisibleToPlayer(whichPoint.handle, this.handle);
    }

    /**
     * 踢除玩家
     */
    public remove(gameResult: playergameresult) {
        RemovePlayer(this.handle, gameResult);
    }

    /**
     * 忽略所有单位的防守职责
     */
    public removeAllGuardPositions() {
        RemoveAllGuardPositions(this.handle);
    }

    /**
     * 允许/禁用 技能 [R]
     */
    public setAbilityAvailable(abilId: number | string, avail: boolean) {
        SetPlayerAbilityAvailable(this.handle, abilId, avail);
    }

    /**
     * 设置联盟状态(指定项目) [R]
     */
    public setAlliance(otherPlayer: MapPlayer, whichAllianceSetting: alliancetype, value: boolean) {
        SetPlayerAlliance(this.handle, otherPlayer.handle, whichAllianceSetting, value);
    }

    /**
     * 显示/隐藏计分屏显示 [R]
     */
    public setOnScoreScreen(flag: boolean) {
        SetPlayerOnScoreScreen(this.handle, flag);
    }

    /**
     * 增加金币
     */
    public addGoldState(value: number) {
        this.addState(PLAYER_STATE_RESOURCE_GOLD, value)
    }

    /**
     * 增加木材
     */
    public addLumberState(value: number) {
        this.addState(PLAYER_STATE_RESOURCE_LUMBER, value)
    }

    /**
     * 拥有指定数量的金币
     */
    public hasGold(value: number): boolean {
        return this.getState(PLAYER_STATE_RESOURCE_GOLD) >= value
    }

    /**
     * 拥有指定数量的木材
     */
    public hasLumber(value: number): boolean {
        return this.getState(PLAYER_STATE_RESOURCE_LUMBER) >= value
    }

    /**
     * 获取金币数量
     */
    public getGold(): number {
        return this.getState(PLAYER_STATE_RESOURCE_GOLD)
    }

    /**
     * 获取木材数量
     */
    public getLumber(): number {
        return this.getState(PLAYER_STATE_RESOURCE_LUMBER)
    }

    /**
     * 添加属性
     */
    public addState(whichPlayerState: playerstate, value: number) {
        this.setState(whichPlayerState, this.getState(whichPlayerState) + value);
    }

    /**
     * 设置属性
     */
    public setState(whichPlayerState: playerstate, value: number) {
        SetPlayerState(this.handle, whichPlayerState, value);
    }

    /**
     * 设置税率 [R]
     */
    public setTaxRate(otherPlayer: MapPlayer, whichResource: playerstate, rate: number) {
        SetPlayerTaxRate(this.handle, otherPlayer.handle, whichResource, rate);
    }

    /**
     * 允许玩家的科技上限
     * @param techId
     * @param maximum
     */
    public setTechMaxAllowed(techId: number | string, maximum: number) {
        SetPlayerTechMaxAllowed(this.handle, techId, maximum);
    }

    /**
     * 设置玩家科技等级
     */
    public setTechResearched(techId: number | string, setToLevel: number) {
        SetPlayerTechResearched(this.handle, techId, setToLevel);
    }

    /**
     * 设置单位所属玩家
     */
    public setUnitsOwner(newOwner: number) {
        SetPlayerUnitsOwner(this.handle, newOwner);
    }

    /**
     * 选取的玩家
     */
    public static fromEnum() {
        return MapPlayer.fromHandle(GetEnumPlayer());
    }

    /**
     * 触发玩家
     */
    public static fromEvent() {
        return MapPlayer.fromHandle(GetTriggerPlayer());
    }

    /**
     * 匹配的玩家
     */
    public static fromFilter() {
        return MapPlayer.fromHandle(GetFilterPlayer());
    }

    /**
     * 从handle获取玩家封装类
     */
    public static fromHandle(handle: player): MapPlayer {
        let obj: MapPlayer = cache[GetHandleId(handle)];
        if (!obj) {
            obj = this.getObject(handle);
            cache[GetHandleId(handle)] = obj;
        }
        return obj;
    }

    /**
     * 从玩家id获取玩家封装类
     */
    public static fromIndex(index: number) {
        return this.fromHandle(Player(index));
    }

    /**
     * 获取本地玩家封装类
     */
    public static fromLocal() {
        return this.fromHandle(GetLocalPlayer());
    }

}

File: scripts/solar/solar-common/w3ts/handles/player.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 4,["9"] = 4,["10"] = 6,["11"] = 6,["12"] = 8,["13"] = 10,["14"] = 10,["15"] = 10,["16"] = 10,["17"] = 15,["18"] = 16,["19"] = 10,["21"] = 10,["22"] = 19,["23"] = 19,["24"] = 19,["26"] = 15,["27"] = 37,["28"] = 38,["29"] = 39,["31"] = 41,["32"] = 37,["33"] = 47,["34"] = 48,["35"] = 49,["36"] = 50,["37"] = 47,["38"] = 56,["39"] = 56,["40"] = 56,["42"] = 57,["43"] = 57,["44"] = 57,["45"] = 57,["46"] = 57,["47"] = 57,["48"] = 57,["49"] = 56,["50"] = 197,["51"] = 198,["52"] = 197,["53"] = 210,["54"] = 211,["55"] = 210,["56"] = 217,["57"] = 218,["58"] = 217,["59"] = 224,["60"] = 225,["61"] = 224,["62"] = 231,["63"] = 232,["64"] = 231,["65"] = 238,["66"] = 239,["67"] = 238,["68"] = 245,["69"] = 246,["70"] = 245,["71"] = 252,["72"] = 253,["73"] = 252,["74"] = 259,["75"] = 260,["76"] = 259,["77"] = 266,["78"] = 267,["79"] = 266,["80"] = 273,["81"] = 274,["82"] = 273,["83"] = 280,["84"] = 281,["85"] = 280,["86"] = 288,["87"] = 289,["88"] = 288,["89"] = 295,["90"] = 296,["91"] = 295,["92"] = 302,["93"] = 303,["94"] = 302,["95"] = 312,["96"] = 313,["97"] = 312,["98"] = 319,["99"] = 320,["100"] = 319,["101"] = 326,["102"] = 327,["103"] = 326,["104"] = 333,["105"] = 334,["106"] = 333,["107"] = 340,["108"] = 341,["109"] = 340,["110"] = 347,["111"] = 348,["112"] = 347,["113"] = 354,["114"] = 355,["115"] = 354,["116"] = 361,["117"] = 362,["118"] = 361,["119"] = 368,["120"] = 369,["121"] = 368,["122"] = 375,["123"] = 376,["124"] = 375,["125"] = 382,["126"] = 383,["127"] = 382,["128"] = 389,["129"] = 390,["130"] = 389,["131"] = 396,["132"] = 397,["133"] = 396,["134"] = 403,["135"] = 404,["136"] = 403,["137"] = 410,["138"] = 411,["139"] = 410,["140"] = 417,["141"] = 418,["142"] = 417,["143"] = 424,["144"] = 425,["145"] = 424,["146"] = 431,["147"] = 432,["148"] = 431,["149"] = 438,["150"] = 439,["151"] = 438,["152"] = 445,["153"] = 446,["154"] = 445,["155"] = 452,["156"] = 453,["157"] = 452,["158"] = 459,["159"] = 460,["160"] = 460,["161"] = 460,["162"] = 460,["163"] = 459,["164"] = 466,["165"] = 467,["166"] = 466,["167"] = 473,["168"] = 474,["169"] = 473,["170"] = 482,["171"] = 483,["172"] = 482,["173"] = 489,["174"] = 490,["175"] = 489,["176"] = 496,["177"] = 497,["178"] = 496,["179"] = 503,["180"] = 504,["181"] = 503,["182"] = 510,["183"] = 511,["184"] = 510,["185"] = 517,["186"] = 518,["187"] = 517,["188"] = 524,["189"] = 525,["190"] = 526,["191"] = 527,["192"] = 528,["194"] = 530,["195"] = 524,["196"] = 536,["197"] = 537,["198"] = 536,["199"] = 543,["200"] = 544,["201"] = 543,["207"] = 27,["209"] = 30,["210"] = 31,["220"] = 84,["222"] = 76,["223"] = 77,["232"] = 91,["241"] = 96,["243"] = 102,["244"] = 103,["254"] = 110,["256"] = 116,["257"] = 117,["266"] = 124,["275"] = 131,["277"] = 137,["278"] = 138,["287"] = 145,["295"] = 152,["303"] = 159,["311"] = 166,["319"] = 173,["327"] = 180,["335"] = 187});
local ____exports = {}
local ____handle = require("solar.solar-common.w3ts.handles.handle")
local Handle = ____handle.Handle
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
local cache = {}
____exports.MapPlayer = __TS__Class()
local MapPlayer = ____exports.MapPlayer
MapPlayer.name = "MapPlayer"
__TS__ClassExtends(MapPlayer, Handle)
function MapPlayer.prototype.____constructor(self, index)
    if Handle:initFromHandle() then
        Handle.prototype.____constructor(self)
    else
        Handle.prototype.____constructor(
            self,
            Player(index)
        )
    end
end
function MapPlayer.prototype.isLocalPlayer(self)
    if GetLocalPlayer() == self.handle then
        return true
    end
    return false
end
function MapPlayer.prototype.getAllUnits(self)
    local g = CreateGroup()
    GroupEnumUnitsOfPlayer(g, self.handle, nil)
    return g
end
function MapPlayer.prototype.displayText(self, text, duration)
    if duration == nil then
        duration = 5
    end
    DisplayTimedTextToPlayer(
        self.handle,
        0,
        0,
        duration,
        text
    )
end
function MapPlayer.prototype.addTechResearched(self, techId, levels)
    AddPlayerTechResearched(self.handle, techId, levels)
end
function MapPlayer.prototype.cacheHeroData(self)
    CachePlayerHeroData(self.handle)
end
function MapPlayer.prototype.compareAlliance(self, otherPlayer, whichAllianceSetting)
    return GetPlayerAlliance(self.handle, otherPlayer.handle, whichAllianceSetting)
end
function MapPlayer.prototype.coordsFogged(self, x, y)
    return IsFoggedToPlayer(x, y, self.handle)
end
function MapPlayer.prototype.coordsMasked(self, x, y)
    return IsMaskedToPlayer(x, y, self.handle)
end
function MapPlayer.prototype.coordsVisible(self, x, y)
    return IsVisibleToPlayer(x, y, self.handle)
end
function MapPlayer.prototype.cripple(self, toWhichPlayers, flag)
    CripplePlayer(self.handle, toWhichPlayers.handle, flag)
end
function MapPlayer.prototype.getScore(self, whichPlayerScore)
    return GetPlayerScore(self.handle, whichPlayerScore)
end
function MapPlayer.prototype.getState(self, whichPlayerState)
    return GetPlayerState(self.handle, whichPlayerState)
end
function MapPlayer.prototype.getStructureCount(self, includeIncomplete)
    return GetPlayerStructureCount(self.handle, includeIncomplete)
end
function MapPlayer.prototype.getTaxRate(self, otherPlayer, whichResource)
    return GetPlayerTaxRate(self.handle, otherPlayer, whichResource)
end
function MapPlayer.prototype.getTechCount(self, techId, specificonly)
    return GetPlayerTechCount(self.handle, techId, specificonly)
end
function MapPlayer.prototype.getTechMaxAllowed(self, techId)
    return GetPlayerTechMaxAllowed(self.handle, techId)
end
function MapPlayer.prototype.getTechResearched(self, techId, specificonly)
    return GetPlayerTechResearched(self.handle, techId, specificonly)
end
function MapPlayer.prototype.getUnitCount(self, includeIncomplete)
    return GetPlayerUnitCount(self.handle, includeIncomplete)
end
function MapPlayer.prototype.getUnitCountByType(self, unitName, includeIncomplete, includeUpgrades)
    return GetPlayerTypedUnitCount(self.handle, unitName, includeIncomplete, includeUpgrades)
end
function MapPlayer.prototype.inForce(self, whichForce)
    return IsPlayerInForce(self.handle, whichForce.handle)
end
function MapPlayer.prototype.isObserver(self)
    return IsPlayerObserver(self.handle)
end
function MapPlayer.prototype.isPlayerAlly(self, otherPlayer)
    return IsPlayerAlly(self.handle, otherPlayer.handle)
end
function MapPlayer.prototype.isPlayerEnemy(self, otherPlayer)
    return IsPlayerEnemy(self.handle, otherPlayer.handle)
end
function MapPlayer.prototype.isRacePrefSet(self, pref)
    return IsPlayerRacePrefSet(self.handle, pref)
end
function MapPlayer.prototype.isSelectable(self)
    return GetPlayerSelectable(self.handle)
end
function MapPlayer.prototype.pointFogged(self, whichPoint)
    return IsLocationFoggedToPlayer(whichPoint.handle, self.handle)
end
function MapPlayer.prototype.pointMasked(self, whichPoint)
    return IsLocationMaskedToPlayer(whichPoint.handle, self.handle)
end
function MapPlayer.prototype.pointVisible(self, whichPoint)
    return IsLocationVisibleToPlayer(whichPoint.handle, self.handle)
end
function MapPlayer.prototype.remove(self, gameResult)
    RemovePlayer(self.handle, gameResult)
end
function MapPlayer.prototype.removeAllGuardPositions(self)
    RemoveAllGuardPositions(self.handle)
end
function MapPlayer.prototype.setAbilityAvailable(self, abilId, avail)
    SetPlayerAbilityAvailable(self.handle, abilId, avail)
end
function MapPlayer.prototype.setAlliance(self, otherPlayer, whichAllianceSetting, value)
    SetPlayerAlliance(self.handle, otherPlayer.handle, whichAllianceSetting, value)
end
function MapPlayer.prototype.setOnScoreScreen(self, flag)
    SetPlayerOnScoreScreen(self.handle, flag)
end
function MapPlayer.prototype.addGoldState(self, value)
    self:addState(PLAYER_STATE_RESOURCE_GOLD, value)
end
function MapPlayer.prototype.addLumberState(self, value)
    self:addState(PLAYER_STATE_RESOURCE_LUMBER, value)
end
function MapPlayer.prototype.hasGold(self, value)
    return self:getState(PLAYER_STATE_RESOURCE_GOLD) >= value
end
function MapPlayer.prototype.hasLumber(self, value)
    return self:getState(PLAYER_STATE_RESOURCE_LUMBER) >= value
end
function MapPlayer.prototype.getGold(self)
    return self:getState(PLAYER_STATE_RESOURCE_GOLD)
end
function MapPlayer.prototype.getLumber(self)
    return self:getState(PLAYER_STATE_RESOURCE_LUMBER)
end
function MapPlayer.prototype.addState(self, whichPlayerState, value)
    self:setState(
        whichPlayerState,
        self:getState(whichPlayerState) + value
    )
end
function MapPlayer.prototype.setState(self, whichPlayerState, value)
    SetPlayerState(self.handle, whichPlayerState, value)
end
function MapPlayer.prototype.setTaxRate(self, otherPlayer, whichResource, rate)
    SetPlayerTaxRate(self.handle, otherPlayer.handle, whichResource, rate)
end
function MapPlayer.prototype.setTechMaxAllowed(self, techId, maximum)
    SetPlayerTechMaxAllowed(self.handle, techId, maximum)
end
function MapPlayer.prototype.setTechResearched(self, techId, setToLevel)
    SetPlayerTechResearched(self.handle, techId, setToLevel)
end
function MapPlayer.prototype.setUnitsOwner(self, newOwner)
    SetPlayerUnitsOwner(self.handle, newOwner)
end
function MapPlayer.fromEnum(self)
    return ____exports.MapPlayer:fromHandle(GetEnumPlayer())
end
function MapPlayer.fromEvent(self)
    return ____exports.MapPlayer:fromHandle(GetTriggerPlayer())
end
function MapPlayer.fromFilter(self)
    return ____exports.MapPlayer:fromHandle(GetFilterPlayer())
end
function MapPlayer.fromHandle(self, handle)
    local obj = cache[GetHandleId(handle)]
    if not obj then
        obj = self:getObject(handle)
        cache[GetHandleId(handle)] = obj
    end
    return obj
end
function MapPlayer.fromIndex(self, index)
    return self:fromHandle(Player(index))
end
function MapPlayer.fromLocal(self)
    return self:fromHandle(GetLocalPlayer())
end
__TS__SetDescriptor(
    MapPlayer.prototype,
    "solarData",
    {
        get = function(self)
            return DataBase:getPlayerSolarData(self.handle)
        end,
        set = function(self, obj)
            DataBase:setDataByHandle("+ply", self.handle, obj)
        end
    },
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "color",
    {
        get = function(self)
            return GetPlayerColor(self.handle)
        end,
        set = function(self, color)
            SetPlayerColor(self.handle, color)
        end
    },
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "controller",
    {get = function(self)
        return GetPlayerController(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "handicap",
    {
        get = function(self)
            return GetPlayerHandicap(self.handle)
        end,
        set = function(self, handicap)
            SetPlayerHandicap(self.handle, handicap)
        end
    },
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "handicapXp",
    {
        get = function(self)
            return GetPlayerHandicapXP(self.handle)
        end,
        set = function(self, handicap)
            SetPlayerHandicapXP(self.handle, handicap)
        end
    },
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "id",
    {get = function(self)
        return GetPlayerId(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "name",
    {
        get = function(self)
            return GetPlayerName(self.handle)
        end,
        set = function(self, value)
            SetPlayerName(self.handle, value)
        end
    },
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "race",
    {get = function(self)
        return GetPlayerRace(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "slotState",
    {get = function(self)
        return GetPlayerSlotState(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "startLocation",
    {get = function(self)
        return GetPlayerStartLocation(self.handle)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "startLocationX",
    {get = function(self)
        return GetStartLocationX(self.startLocation)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "startLocationY",
    {get = function(self)
        return GetStartLocationY(self.startLocation)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "startLocationPoint",
    {get = function(self)
        return GetStartLocationLoc(self.startLocation)
    end},
    true
)
__TS__SetDescriptor(
    MapPlayer.prototype,
    "team",
    {get = function(self)
        return GetPlayerTeam(self.handle)
    end},
    true
)
return ____exports
